domain/network/state/space.go

```diff
@@ -0,0 +1,292 @@
+}
+
+const retrieveSpacesStmt = `
+SELECT     
+    space.uuid                           AS &Space.uuid,
+    space.name                           AS &Space.name,
+    provider_space.provider_id           AS &Space.provider_id,
+    subnet.uuid                          AS &Space.subnet_uuid,
+    subnet.cidr                          AS &Space.subnet_cidr,
+    subnet.vlan_tag                      AS &Space.vlan_tag,
+    provider_subnet.provider_id          AS &Space.subnet_provider_id,
+    provider_network.provider_network_id AS &Space.subnet_provider_network_id,
+    availability_zone.name               AS &Space.subnet_az
+FROM space 
+    LEFT JOIN provider_space
+    ON space.uuid = provider_space.space_uuid
+    LEFT JOIN subnet   
+    ON space.uuid = subnet.space_uuid
+    LEFT JOIN provider_subnet
+    ON subnet.uuid = provider_subnet.subnet_uuid
+    LEFT JOIN provider_network_subnet
+    ON subnet.uuid = provider_network_subnet.subnet_uuid
+    LEFT JOIN provider_network
+    ON provider_network_subnet.provider_network_uuid = provider_network.uuid
+    LEFT JOIN availability_zone_subnet
+    ON availability_zone_subnet.subnet_uuid = subnet.uuid
+    LEFT JOIN availability_zone
+    ON availability_zone_subnet.availability_zone_uuid = availability_zone.uuid`
+
+// GetSpace returns the space by UUID.
+func (st *State) GetSpace(
+	ctx context.Context,
+	uuid string,
+) (*network.SpaceInfo, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Append the space uuid condition to the query only if it's passed to the function.
+	q := retrieveSpacesStmt + " WHERE space.uuid = $M.id;"
+
+	s, err := sqlair.Prepare(q, Space{}, sqlair.M{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", q)
+	}
+
+	var rows Spaces
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{"id": uuid}).GetAll(&rows))
+	}); err != nil {
+		return nil, errors.Annotate(err, "querying spaces")
+	}
+
+	if len(rows) == 0 {
+		return nil, errors.NotFoundf("space %q", uuid)
+	}
+
+	return &rows.ToSpaceInfos()[0], nil
+}
+
+// GetSpace returns the space by name.
+func (st *State) GetSpaceByName(
+	ctx context.Context,
+	name string,
+) (*network.SpaceInfo, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Append the space.name condition to the query.
+	q := retrieveSpacesStmt + " WHERE space.name = $M.name;"
+
+	s, err := sqlair.Prepare(q, Space{}, sqlair.M{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", q)
+	}
+
+	var rows Spaces
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{"name": name}).GetAll(&rows))
+	}); err != nil {
+		return nil, errors.Annotate(err, "querying spaces by name")
+	}
+
+	if len(rows) == 0 {
+		return nil, errors.NotFoundf("space name %q", name)
+	}
+
+	return &rows.ToSpaceInfos()[0], nil
```

Comment:
Can we not use `.Get` here? It feels like you could detect the `sql.ErrNoRows` instead of empty slice.